Add hipBLASLt examples (#311)
amd/rocm-examples: Libraries/hipBLASLt/ext_op_amax/CMakeLists.txt

# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

set(example_name hipblaslt_ext_op_amax)

cmake_minimum_required(VERSION 3.21 FATAL_ERROR)
project(${example_name} LANGUAGES CXX)

if(GPU_RUNTIME STREQUAL "CUDA")
    message(STATUS "hipBLASLt examples do not support the CUDA runtime")
    return()
endif()

> enable_language(${GPU_RUNTIME})
> set(CMAKE_${GPU_RUNTIME}_STANDARD 17)
> set(CMAKE_${GPU_RUNTIME}_EXTENSIONS OFF)
> set(CMAKE_${GPU_RUNTIME}_STANDARD_REQUIRED ON)
> 
> if(CMAKE_SYSTEM_NAME MATCHES "Windows")
>     message(STATUS "hipBLASLt examples are only available on Linux")
>     return()
> else()
>     set(ROCM_ROOT
>         "/opt/rocm"
>         CACHE PATH
>         "Root directory of the ROCm installation"
>     )
> endif()
> 
> list(APPEND CMAKE_PREFIX_PATH "${ROCM_ROOT}")
> 
> find_package(hipblaslt REQUIRED)
> 
> add_executable(${example_name} main.cpp)
> # Make example runnable using ctest
> add_test(NAME ${example_name} COMMAND ${example_name})
> 
> # Link to example library
> target_link_libraries(${example_name} PRIVATE roc::hipblaslt)
> target_compile_definitions(${example_name} PRIVATE ROCM_USE_FLOAT16)

Reviewer: If you want to get rid of those annoying deprecation warnings, you can do this:

```suggestion
include("${CMAKE_CURRENT_LIST_DIR}/../../../Common/HipPlatform.cmake")
select_gpu_language()
enable_language(${ROCM_EXAMPLES_GPU_LANGUAGE})
select_hip_platform()

if(CMAKE_SYSTEM_NAME MATCHES "Windows")
    message(STATUS "hipBLASLt examples are only available on Linux")
    return()
else()
    set(ROCM_ROOT
        "/opt/rocm"
        CACHE PATH
        "Root directory of the ROCm installation"
    )
endif()

list(APPEND CMAKE_PREFIX_PATH "${ROCM_ROOT}")

find_package(hipblaslt REQUIRED)

add_executable(${example_name} main.cpp)
# Make example runnable using ctest
add_test(NAME ${example_name} COMMAND ${example_name})

set_target_properties(${example_name} PROPERTIES CXX_EXTENSIONS OFF)
target_compile_features(${example_name} PRIVATE cxx_std_17)

# Link to example library
target_link_libraries(${example_name} PRIVATE roc::hipblaslt)
target_compile_definitions(${example_name} PRIVATE ROCM_USE_FLOAT16)
```

target_include_directories(
    ${example_name}

>     PRIVATE "../../../Common" "../../../External"

Reviewer: ```suggestion
    PRIVATE "../../../Common"
```

Unless you are actually using one of the external libraries, of course.

)

> set_source_files_properties(main.cpp PROPERTIES LANGUAGE ${GPU_RUNTIME})

Reviewer: ```suggestion
```

From the looks of it, this is a pure C++ code, so we don't need to go the extra steps through HIP.

Author: This is more for consistency (since the makefile just uses hipcc) so I think better to leave this one as is

install(TARGETS ${example_name})
if(CMAKE_SYSTEM_NAME MATCHES "Windows")
    install(IMPORTED_RUNTIME_ARTIFACTS roc::hipblaslt)

> endif()

Reviewer: Isn't Windows disabled?
